libcontainer/cgroups/systemd/v1.go

```diff
@@ -332,6 +331,34 @@ func (m *legacyManager) GetStats() (*cgroups.Stats, error) {
 	return stats, nil
 }
 
+// hasDeviceRules checks cgroup device rules. Returns true if any rules
+// other than "allow all" exist. Used by freezeBeforeSet.
+func (m *legacyManager) hasDeviceRules() bool {
+	const allowAll = "a *:* rwm\n"
+
+	path := m.Path("devices")
+	// XXX: if Apply has not been called, m.paths are not initialized, so
+	// we have to figure the path. TODO: remove once fixed.
+	if path == "" {
+		var err error
+		path, err = getSubsystemPath(m.cgroups, "devices")
+		if err != nil {
+			// Should not happen. Assume we have rules.
+			return true
+		}
+	}
+	list, err := cgroups.ReadFile(path, "devices.list")
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return false
+		}
+		// Should not happen. Log the error, assume we do have rules.
+		logrus.Debug(err)
+		return true
+	}
+	return list != allowAll
```

Comment:
about the `\n` in `allowAll` should we use string.TrimSpace or string.Contains to essentially ignore the newline? (or is this `\n` an essential part of the check?)

Reply:
We can use `strings.TrimSpace` for sure, but in this particular case we're not parsing the string, we're merely checking that the entirety of file contents is this string plus a newline. As this is a kernel file and not some user-provided contents we can be sure there are no extra whitespace or some such.